pkg/auth/manager/oidcescape/oidcescape.go

```diff
@@ -0,0 +1,272 @@
+	rpc "github.com/cs3org/go-cs3apis/cs3/rpc/v1beta1"
+	"github.com/cs3org/reva/pkg/appctx"
+	"github.com/cs3org/reva/pkg/auth"
+	"github.com/cs3org/reva/pkg/auth/manager/registry"
+	"github.com/cs3org/reva/pkg/auth/scope"
+	"github.com/cs3org/reva/pkg/rgrpc/todo/pool"
+	"github.com/cs3org/reva/pkg/rhttp"
+	"github.com/mitchellh/mapstructure"
+	"github.com/pkg/errors"
+	"golang.org/x/oauth2"
+)
+
+func init() {
+	registry.Register("oidcescape", New)
+}
+
+type mgr struct {
+	provider *oidc.Provider // cached on first request
+	c        *config
+	iamUsers map[string]*iamUser
+}
+
+type config struct {
+	Insecure        bool   `mapstructure:"insecure" docs:"false;Whether to skip certificate checks when sending requests."`
+	Issuer          string `mapstructure:"issuer" docs:";The issuer of the OIDC token."`
+	IDClaim         string `mapstructure:"id_claim" docs:"sub;The claim containing the ID of the user."`
+	UIDClaim        string `mapstructure:"uid_claim" docs:";The claim containing the UID of the user."`
+	GIDClaim        string `mapstructure:"gid_claim" docs:";The claim containing the GID of the user."`
+	UserProviderSvc string `mapstructure:"userprovidersvc" docs:";The endpoint at which the GRPC userprovider is exposed."`
+	Users           string `mapstructure:"users" docs:"; The IAM users mapping file path"`
+}
+
+type iamUser struct {
+	OpaqueID string `mapstructure:"opaque_id" json:"opaque_id"`
+	Sub      string `mapstructure:"sub" json:"sub"`
+}
+
+func (c *config) init() {
+	if c.IDClaim == "" {
+		c.IDClaim = "sub"
+	}
+	if c.Users == "" {
+		c.Users = "/etc/revad/iam-escape-users.json"
+	}
+}
+
+func parseConfig(m map[string]interface{}) (*config, error) {
+	c := &config{}
+	if err := mapstructure.Decode(m, c); err != nil {
+		err = errors.Wrap(err, "error decoding conf")
+		return nil, err
+	}
+	return c, nil
+}
+
+// New returns an auth manager.
+func New(m map[string]interface{}) (auth.Manager, error) {
+	manager := &mgr{}
+	err := manager.Configure(m)
+	if err != nil {
+		return nil, err
+	}
+	return manager, nil
+}
+
+func (am *mgr) Configure(m map[string]interface{}) error {
+	c, err := parseConfig(m)
+	if err != nil {
+		return err
+	}
+	c.init()
+	am.c = c
+
+	am.iamUsers = map[string]*iamUser{}
+	f, err := ioutil.ReadFile(c.Users)
+	if err != nil {
+		return fmt.Errorf("oidcescape: error reading iam users file: +%v", err)
+	}
+
+	iamUsers := []*iamUser{}
+
+	err = json.Unmarshal(f, &iamUsers)
+	if err != nil {
+		return fmt.Errorf("oidcescape: error unmarshalling iam users file: +%v", err)
+	}
+
+	for _, u := range iamUsers {
+		am.iamUsers[u.Sub] = u
+	}
+
+	return nil
+}
+
+// Authenticate clientID would be empty as we only need to validate the clientSecret variable
+// which contains the access token that we can use to contact the UserInfo endpoint
+// and get the user claims.
+func (am *mgr) Authenticate(ctx context.Context, clientID, clientSecret string) (*user.User, map[string]*authpb.Scope, error) {
```

Comment:
This code is largely identical to the `Authenticate()` function in `pkg/auth/manager/oidc/oidc.go`, suggest to factorize it out, especially as the numerous TODOs from Hugo should hopefully be addressed in a single place for the benefit of both.

Reply:
(as discussed we keep it separate, after experimenting we may well merge this to the oidc driver)